components/scream/src/physics/shoc/tests/shoc_diag_third_tests.cpp

```diff
@@ -0,0 +1,244 @@
+      zt_grid[n] = 0.5*(zi_grid[n]+zi_grid[n+1]);
+      tke[n] = 1.5*w_sec[n];
+      dz_zt[n] = zi_grid[n] - zi_grid[n+1];
+      if (n == 0){
+        dz_zi[n] = 0;
+      }
+      else{
+        dz_zi[n] = zt_grid[n-1] - zt_grid[n];
+      }
+    }
+    // set upper condition for dz_zi
+    dz_zi[nlevi-1] = zt_grid[nlev-1];
+
+    // Initialize data structure for bridging to F90
+    SHOCDiagThirdMomData SDS(shcol, nlev, nlevi);
+
+    // Test that the inputs are reasonable.
+    // For this test shcol MUST be at least 2
+    REQUIRE( (SDS.shcol() == shcol && SDS.nlev() == nlev && SDS.nlevi() == nlevi) );
+    REQUIRE(SDS.nlevi() == SDS.nlev()+1);
+    
+    // Load up the new data
+    for(Int s = 0; s < shcol; ++s) { 
+      // Fill in test data on zt_grid.     
+      for(Int n = 0; n < nlev; ++n) {
+        const auto offset = n + s * nlev;
+	
+        SDS.w_sec[offset] = w_sec[n];
+        SDS.dz_zt[offset] = dz_zt[n];
+        SDS.zt_grid[offset] = zt_grid[n];
+        SDS.tke[offset] = tke[n];
+	
+        SDS.isotropy[offset] = isotropy[n];
+        SDS.brunt[offset] = brunt[n];
+        SDS.thetal[offset] = thetal[n];
+        SDS.wthv_sec[offset] = wthv_sec[n];	
+	
+      }
+
+      // Fill in test data on zi_grid.     
+      for(Int n = 0; n < nlevi; ++n) {
+        const auto offset = n + s * nlevi;
+	
+        SDS.dz_zi[offset] = dz_zi[n];
+        SDS.zi_grid[offset] = zi_grid[n];
+        SDS.thl_sec[offset] = thl_sec[n];
+        SDS.qw_sec[offset] = qw_sec[n];
+        SDS.qwthl_sec[offset] = qwthl_sec[n];
+        SDS.wthl_sec[offset] = wthl_sec[n];
+	
+      }
+    }      
+
+    // Check that the inputs make sense
+    // Load up the new data
+    for(Int s = 0; s < shcol; ++s) { 
+      // Fill in test data on zt_grid.    
+      for(Int n = 0; n < nlev; ++n) {
+        const auto offset = n + s * nlev;
+	
+        REQUIRE(SDS.w_sec[offset] >= 0);
+	       // for this test make sure w_sec <= 1
+        REQUIRE(SDS.w_sec[offset] <= 1);
+	
+        REQUIRE(SDS.dz_zt[offset] > 0);
+        REQUIRE(SDS.zt_grid[offset] > 0);
+        REQUIRE(SDS.tke[offset] > 0);
+        REQUIRE(SDS.isotropy[offset] >= 0);
+        REQUIRE(SDS.thetal[offset] >= 0);
+      }
+      
+      for(Int n = 0; n < nlevi; ++n) {
+        const auto offset = n + s * nlevi;
+	
+        REQUIRE(SDS.dz_zi[offset] >= 0);
+        REQUIRE(SDS.zi_grid[offset] >= 0);
+        REQUIRE(SDS.thl_sec[offset] >= 0);
+        REQUIRE(SDS.qw_sec[offset] >= 0);
+	
+      }    
+      
+    }
+
+    // Call the fortran implementation
+    diag_third_shoc_moments(SDS);
+    
+    // Check to make sure there is at least one
+    //  positive w3 value for convective boundary layer
+    bool is_skew;
+    // Verify that boundary points are zero
+    for(Int s = 0; s < shcol; ++s) { 
+      is_skew = false; // Initialize
+      for(Int n = 0; n < nlevi; ++n) {
+        const auto offset = n + s * nlevi;
+
+	       // For this test make sure w3 has been clipped.
```

Comment:
Some strange indentation going on here through line 212. Looks like some tabs are getting mixed-in too.